### PR TITLE
Test Failures | "TypeError: expected bytes, numpy.ndarray found" error fix

### PR DESCRIPTION
**Pull request by MounikaBattu17**
- [x] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/nidaqmx-python/blob/master/CONTRIBUTING.md).

### What does this Pull Request accomplish?

This PR fixes the error **"TypeError: expected bytes, numpy.ndarray found"** while running tests.
- Currently `write_digital_lines`,` write_digital_u8`, `write_raw` request takes in `bytes` instead of `ndarray` and same `read_digital_lines`, `read_digital_u8`,  `read_raw` reads `bytes` from the channel as per the requst and response in `nidqmx.pb2.pyi` file.
   - Updated `interpreter_helpers.py` to type cast `write_array` to bytes for these write functions.
     `write_array=bytes(write_array)`
  - Updated `read_digital_lines`, `read_digital_u8`,  `read_raw` to convert bytes back to ndarray in `_grpc_interpreter.mako` file
        `if isinstance(grpc_array, bytes):`
        `grpc_array = numpy.frombuffer(grpc_array, dtype=numpy.uint8)`
### Why should this Pull Request be merged?

This PR fixes the error **"TypeError: expected bytes, numpy.ndarray found"** while running tests.

### What testing has been done?

Ran `poetry run pytest`. This issue tests passed.

### CODE REVIEW

**Review by zhindes — APPROVED** [review]

File: generated/nidaqmx/_grpc_interpreter.py [hunk 1]
```diff
@@ -3384,9 +3385,13 @@ def write_to_teds_from_file(
                 basic_teds_options_raw=basic_teds_options))
 
 def _assign_numpy_array(numpy_array, grpc_array):
-    """ Assigns the grpc array to the numpy array, while still maintaining the 
-    original shape of the numpy array. 
+    """ 
+    Assigns the grpc array to the numpy array, while still maintaining the 
+    original shape of the numpy array.
+    Checks if grpc_array is an instance of bytes, then converts grpc_array to a 1D array. 
     """
```
Comment (bkeryan) [inline generated/nidaqmx/_grpc_interpreter.py:3392]: https://peps.python.org/pep-0257/#multi-line-docstrings

> Multi-line docstrings consist of a summary line just like a one-line docstring, followed by a blank line, followed by a more elaborate description. The summary line may be used by automatic indexing tools; it is important that it fits on one line and is separated from the rest of the docstring by a blank line.

File: generated/nidaqmx/_grpc_interpreter.py [hunk 2]
```diff
@@ -3356,15 +3357,15 @@ def write_digital_u8(
             grpc_types.WriteDigitalU8Request(
                 task=task, num_samps_per_chan=num_samps_per_chan,
                 auto_start=auto_start, timeout=timeout,
-                data_layout_raw=data_layout, write_array=write_array))
+                data_layout_raw=data_layout, write_array=bytes(write_array)))
```
Comment (bkeryan) [inline generated/nidaqmx/_grpc_interpreter.py:3360]: This does not validate that write_array has the correct dtype. If the array has the wrong dtype, this writes the wrong data.

LibraryInterpreter validates the array's dtype by specifying it on the ndpointer_wrapper.
Example code:
```
import nidaqmx
import numpy
from nidaqmx.constants import LineGrouping
from nidaqmx.stream_writers import DigitalSingleChannelWriter

with nidaqmx.Task() as task:
    task.do_channels.add_do_chan(
        "Dev1/port1/line0:3", line_grouping=LineGrouping.CHAN_FOR_ALL_LINES
    )

    writer = DigitalSingleChannelWriter(task.out_stream, auto_start=True)
    write_array = numpy.array([1, 2, 4, 8], dtype=numpy.uint8)
    print(writer.write_many_sample_port_byte(write_array))
```

With the correct dtype of uint8, it writes 4 bytes.

With the incorrect dtype of uint32, it raises an exception:
```
PS D:\dev\nidaqmx-python> poetry run python .\examples\do_sw_timed.py
Traceback (most recent call last):
  File "D:\dev\nidaqmx-python\examples\do_sw_timed.py", line 14, in <module>
    print(writer.write_many_sample_port_byte(write_array))
  File "D:\dev\nidaqmx-python\generated\nidaqmx\stream_writers.py", line 836, in write_many_sample_port_byte
    return self._interpreter.write_digital_u8(
  File "D:\dev\nidaqmx-python\generated\nidaqmx\_library_interpreter.py", line 5850, in write_digital_u8
    error_code = cfunc(
ctypes.ArgumentError: argument 6: <class 'TypeError'>: array must have data type uint8
```

GrpcStubInterpreter is using `bytes(write_array)`. With a dtype of uint32, `print(bytes(write_array))` prints a lot of zero bytes:
```
b'\x01\x00\x00\x00\x02\x00\x00\x00\x04\x00\x00\x00\x08\x00\x00\x00'
```

NI IO Trace shows that this writes the zero bytes:
```
177681.  DAQmxWriteDigitalU8 (0x00007FFBD523CD28, 4 (0x4), True, 10.000000 (1.000000E+01), "DAQmx_Val_GroupByChannel", {0x01,0x00,...}, 4 (0x4), NULL)
Process ID: 0x0000246C         Thread ID: 0x00003F98
Start Time: 14:29:39.2682      Call Duration 00:00:00.0000
Status: 0 (0x0)
```
image

**Review by bkeryan — CHANGES_REQUESTED** [review]

File: generated/nidaqmx/_grpc_interpreter.py (hunk 2 above)
Comment (bkeryan) [reply generated/nidaqmx/_grpc_interpreter.py:3360]: This affects a lot more than these three functions, so I filed an internal bug: AB#2390189: GrpcStubInterpreter reads/writes invalid data when NumPy array has wrong dtype

File: generated/nidaqmx/_grpc_interpreter.py [hunk 3]
```diff
@@ -3384,9 +3385,13 @@ def write_to_teds_from_file(
...
+    """ 
+    Assigns the grpc array to the numpy array, while still maintaining the 
+    original shape of the numpy array.
+    Checks if grpc_array is an instance of bytes, then converts grpc_array to a 1D array. 
     """
     grpc_array_size = len(grpc_array)
     assert numpy_array.size >= grpc_array_size
+    if isinstance(grpc_array, bytes):
+        grpc_array = numpy.frombuffer(grpc_array, dtype=numpy.uint8)
     numpy_array.flat[:grpc_array_size] = grpc_array
```
[the remark this answers is not in the document]
Comment (MounikaBattu17) [reply generated/nidaqmx/_grpc_interpreter.py]: Updated.

File: generated/nidaqmx/_grpc_interpreter.py (hunk 1 above)
Comment (MounikaBattu17) [reply generated/nidaqmx/_grpc_interpreter.py]: Restructured the multiline docstring.

Comment (bkeryan) [reply generated/nidaqmx/_grpc_interpreter.py]: There should be a one line summary, then a blank line, then additional paragraphs separated by blank lines.

**Review by bkeryan — APPROVED** [review]
Approved with suggestion:
- Use the correct docstring format

File: generated/nidaqmx/_grpc_interpreter.py (hunk 1 above)
Comment (MounikaBattu17) [reply generated/nidaqmx/_grpc_interpreter.py]: Thank you for the example snippet. I have updated the docstring to this format.

**Pull request merged**